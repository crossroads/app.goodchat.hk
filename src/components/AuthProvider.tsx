import React, { useState } from "react";
import AuthContext from "../context/AuthContext";

interface Props {
  initialAuthState?: boolean;
}
const AuthProvider: React.FC<Props> = ({
  children,
  initialAuthState = false,
}) => {
  const [isAuthenticated, setIsAuthenticated] = useState(initialAuthState);

  return (
    <AuthContext.Provider value={{ isAuthenticated, setIsAuthenticated }}>
      {children}
    </AuthContext.Provider>
  );
};

export default AuthProvider;
