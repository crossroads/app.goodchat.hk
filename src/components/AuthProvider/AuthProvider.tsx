import React, { useState } from "react";
import AuthContext from "context/AuthContext";
import { AUTHENTICATED } from "config/localStorageKeys";

export const computeAuthState = (initialAuthState: boolean) => {
  return Boolean(localStorage.getItem(AUTHENTICATED) ?? initialAuthState);
};
interface Props {
  initialAuthState?: boolean;
}
const AuthProvider: React.FC<Props> = ({
  children,
  initialAuthState = false,
}) => {
  const [isAuthenticated, setIsAuthenticated] = useState(() =>
    computeAuthState(initialAuthState)
  );

  return (
    <AuthContext.Provider value={{ isAuthenticated, setIsAuthenticated }}>
      {children}
    </AuthContext.Provider>
  );
};

export default AuthProvider;
