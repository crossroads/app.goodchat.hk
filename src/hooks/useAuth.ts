import { useContext } from "react";
import AuthContext from "context/AuthContext";

export interface Auth {
  isAuthenticated: boolean;
  setIsAuthenticated: React.Dispatch<React.SetStateAction<boolean>>;
}
const useAuth = (): Auth => useContext(AuthContext);

export default useAuth;
