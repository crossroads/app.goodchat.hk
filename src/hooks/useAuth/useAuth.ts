import { useContext } from "react";
import AuthContext from "context/AuthContext";
import AuthenticationService from "lib/services/AuthenticationService/AuthenticationService";

export interface Auth {
  isAuthenticated: boolean;
  login: (pin: string) => void;
  logout: () => void;
}
const useAuth = (): Auth => {
  const { isAuthenticated, setIsAuthenticated } = useContext(AuthContext);

  const login = async (pin: string) => {
    await AuthenticationService.authenticate(pin);
    setIsAuthenticated(true);
  };

  const logout = () => {
    AuthenticationService.logout();
    setIsAuthenticated(false);
  };

  return {
    isAuthenticated,
    login,
    logout,
  };
};

export default useAuth;
