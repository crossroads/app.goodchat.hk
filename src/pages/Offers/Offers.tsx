import React from "react";
import { IonHeader, IonPage, IonTitle, IonToolbar } from "@ionic/react";

const Offers = () => {
  return (
    <IonPage>
      <IonHeader>
        <IonToolbar>
          <IonTitle>Offers</IonTitle>
        </IonToolbar>
      </IonHeader>
    </IonPage>
  );
};

export default Offers;
