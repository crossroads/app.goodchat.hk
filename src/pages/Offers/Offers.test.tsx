import React from "react";
import { render, screen } from "@testing-library/react";
import Offers from "pages/Offers/Offers";
import userEvent from "@testing-library/user-event";
import AuthProvider from "components/AuthProvider/AuthProvider";
import * as UseAuthModule from "hooks/useAuth/useAuth";

test("renders correctly", () => {
  const { container } = render(<Offers />);
  expect(container).toMatchSnapshot();
});

test("clicking log out button should call the logout function", () => {
  const mockLogout = jest.fn();
  const mockUseAuth = jest.spyOn(UseAuthModule, "default").mockReturnValue({
    isAuthenticated: true,
    login: jest.fn(),
    logout: mockLogout,
  });

  render(
    <AuthProvider initialAuthState={true}>
      <Offers />
    </AuthProvider>
  );

  const logoutButton = screen.getByText(/log out/i);
  userEvent.click(logoutButton);

  expect(mockLogout).toHaveBeenCalledTimes(1);

  mockUseAuth.mockRestore();
});
