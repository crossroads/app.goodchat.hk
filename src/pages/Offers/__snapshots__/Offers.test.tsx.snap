// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    class="ion-page"
  >
    <ion-header>
      <ion-toolbar>
        <ion-title>
          Offers
        </ion-title>
        <ion-buttons
          slot="end"
        >
          <ion-button>
            Log out
          </ion-button>
        </ion-buttons>
      </ion-toolbar>
    </ion-header>
  </div>
</div>
`;
